### DISCUSSION TITLE
CustomPropertyDrawer for Keys

### DISCUSSION THREAD

**Discussion by jasonlawless** [upscalebaby/generic-serializable-dictionary#11] [Q&A] [answered]
So I took your advice on creating a `CustomPropertyDrawer` for my custom struct `ChunkID`.
I created it but it's throwing me a fit.

```csharp
using System.Reflection;
using UnityEngine;
using UnityEditor;

[CustomPropertyDrawer(typeof(ChunkID))]
public class ChunkIDPropertyDrawer : PropertyDrawer
{
    public override void OnGUI(Rect position, SerializedProperty property, GUIContent label)
    {
        var targetObject = property.serializedObject.targetObject;
        var classType = targetObject.GetType();
        var field = classType.GetField(property.propertyPath, BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.DeclaredOnly | BindingFlags.Public);

        EditorGUI.LabelField(position, property.propertyPath);
    }
}
```

Here is what I got so far. And I have created a `public ChunkID testID` in the same SO just to make sure it works. And with my property drawer it does, but when it comes to it being a `Key` in the dictionary it doesn't work.
I think it has to do with `property.propertyPath` because for the public field it's just `testID`, but for your Key field it's `chunks.list.Array.data[0].Key`, which reflection can't find the field.

Can you recommend me the best way to customize the property drawer for my struct to work with your key? 
Perhaps I'm going about this all wrong?

I just want the property to display a string value instead of nesting a breakdown of the fields.

**Comment by upscalebaby** [comment]
Yes I've had quite some pains with designing property drawers myself.  Their API almost encourages you to make stuff that breaks easily.

So you want to just fetch a field and display that instead of all the fields (including expand arrows) from your class, did I understand you correctly?

Have you tried using the serialized objects iterator?

**Reply by jasonlawless** [reply]
Yes I just want to display a field. It simply will take the ChunkID instance and do a ToString on it, that’s it. I’m afraid I do not know what you mean about the iterator. Can you give me an example please?

**Reply by upscalebaby** [reply]
Sure thing, check this out: https://docs.unity3d.com/2020.2/Documentation/ScriptReference/SerializedObject.GetIterator.html

You get back an iterator that you can use to iterate the properties of your serialized object, this way the drawer can become "context agnostic" so it doesn't break when embedded inside other contexts (e.g. nested inside some other class)

Edit: so to be clear, there's no need for either reflection nor getting the target object - you can just work from the serialized properties. Holler if anything is unclear

**Reply by upscalebaby** [reply]
@jasonlawless Did you get it working? Otherwise I can post some source code that might help

**Comment by jasonlawless** [comment]
Sorry for delay, got sidetracked. And thank you for your follow up.
So I been messing around and honestly am now stuck.

The reason why I used reflection via targetObject before was simply because my struct did NOT inherit from `UnityEngine.Object` so none of the `SerializedProperty.whateverValue` would work for this. 

Here is what I managed to tap into so far:
```csharp

public override void OnGUI(Rect position, SerializedProperty property, GUIContent label)
    {
        //base.OnGUI(position, property, label);
        SerializedProperty prop = property.serializedObject.GetIterator();
        prop.Next(true);

        do
        {
            if (prop.type == "GenericDictionary`2")
            {
                Debug.Log("GenericDictionary");
                prop.Next(true); // KeyValuePair
                Debug.Log($"{prop.type} | {fieldInfo.FieldType}");
                prop.Next(true); // Array
                Debug.Log($"Value type: {prop.arrayElementType} | {fieldInfo.FieldType}");
                prop.Next(true); // ArraySize
                Debug.Log($"Dictionary Size: {prop.intValue} | {fieldInfo.FieldType}");
            }
        }
        while (prop.Next(false));
    }
```
Console result:
![image](https://user-images.githubusercontent.com/25869800/109206592-e8b7fa80-775c-11eb-88ab-ef0797027c96.png)

But I don't know how to go farther from here. If I am overriding a property that is strictly `ChunkID` it's as simple as getting the target object and casting it to string to get the label that I want. But with `GenericDictionary` wrapper, I'm in unfamiliar territory.

I simply need to get to each `Key` of type `ChunkID` and just `ToString` the bad boy. No need for me to have it display the child fields `X,Y,Z`. If you can help me navigate through this, that'd be awesome!

This script would need to to be contingent for the bare `ChunkID` field and for the `GenericDictionary<ChunkID, otherValue>` dictionary.
I don't think I want the `Value` to be converted this way because I would like to leave that edittable in case a future use case demands it.

**Reply by upscalebaby** [reply]
> Sorry for delay, got sidetracked. And thank you for your follow up.
> So I been messing around and honestly am now stuck.

Haha no worries I was just curious :)

> The reason why I used reflection via targetObject before was simply because my struct did NOT inherit from `UnityEngine.Object` so none of the `SerializedProperty.whateverValue` would work for this.

Do note that any serializable class or struct is serialized by Unity (no UnityEngine.Object inheritance required!):

```cs
[System.Serializable]
public struct CustomStruct
{
    public string name;
    public Vector3 someVector;
    public float someFloat;
}
```

```cs
[CustomPropertyDrawer(typeof(CustomStruct))]
public class CustomClassDrawer : PropertyDrawer
{
    public override void OnGUI(Rect pos, SerializedProperty property, GUIContent label)
    {
        var text = property.FindPropertyRelative("name");
        EditorGUI.PropertyField(pos, text);
    }

    public override float GetPropertyHeight(SerializedProperty property, GUIContent label)
    {
        return EditorGUIUtility.singleLineHeight;
    }
}
```

This works, it would also work with the iterator of course, but I'm not sure how much that helps in this case though.

The issue is that since the property is now a property of another serialized object (the dictionary) the targetObject refers to the dictionary, not the ChunkID (which you figured out already).

I believe I've actually been in a very similar situation to you with this very project before. I tried using reflection in the GenericDictionaryProperty drawer to do the key collision checking. I did not get that working exactly how I wanted so I figured I was okay with just serializing this data as a bool and the problem disappeared.

There are of course reasons why that might not be acceptable but I'm just throwing it out there for consideration.

There are many resources with both discussions and of helper classes that solve issues like this. I'm not sure of what's the best approach given the Unity we have today - but here are the resources I looked at last time I was in a similar position as you:

http://sketchyventures.com/2015/08/07/unity-tip-getting-the-actual-object-from-a-custom-property-drawer/
https://forum.unity.com/threads/get-a-general-object-value-from-serializedproperty.327098/

Edit: I might take another stab at this later to get some closure on this subject, I'll let you know if I find anything. If you find something and have time don't hesitate to keep me posted!

**Reply by jasonlawless** [reply]
Thank you for the resources, I read them and although it did give a clearer understanding of Unity's serialization, I still can't figure out how to solve this.

**Comment by upscalebaby** [comment]
Hi again @jasonlawless I've spent some more time researching this. TLDR Unitys api is still not fit to handle this use case (from what I can tell) - but workarounds that aren't too much of a hassle do exist.

Quote from the thread I linked before:

> Unity actually does have an implementation of this (almost). The UnityEditor.ScriptAttributeUtility class has a GetFieldInfoFromPropertyPath method which does the same thing we're all doing - splits the given property path by "." and goes through each part to find the field info and handle arrays.
>
>There are two sad parts to that:
>- It's an internal class so we can't access it without reflection.
>- It doesn't return the target object we want, only the final FieldInfo so we would need to re->implement the entire thing anyway.
>So they've done like 90% of the work we need but won't allow us to make use of it ...

I did quick check in my editor and this class is still internal (and still most likely unusable for this anyways).

The two ways that it's still possible to get this working with is by using either one of the utility classes by either Kybernetik or lordofduct in the thread: https://forum.unity.com/threads/get-a-general-object-value-from-serializedproperty.327098/

Here I copy pasted the three relevant methods from lordofducts class just to give it a quick spin (and it does work):

```cs
using UnityEngine;
using UnityEditor;
using System.Reflection;

[CustomPropertyDrawer(typeof(TestStruct))]
public class TestStructDrawer : PropertyDrawer
{
    public override void OnGUI(Rect position, SerializedProperty property, GUIContent label)
    {
        var target = GetTargetObjectOfProperty(property);
        EditorGUI.LabelField(position, target.ToString());
    }

    public static object GetTargetObjectOfProperty(SerializedProperty prop)
    {
        if (prop == null) return null;

        var path = prop.propertyPath.Replace(".Array.data[", "[");
        object obj = prop.serializedObject.targetObject;
        var elements = path.Split('.');
        foreach (var element in elements)
        {
            if (element.Contains("["))
            {
                var elementName = element.Substring(0, element.IndexOf("["));
                var index = System.Convert.ToInt32(element.Substring(element.IndexOf("[")).Replace("[", "").Replace("]", ""));
                obj = GetValue_Imp(obj, elementName, index);
            }
            else
            {
                obj = GetValue_Imp(obj, element);
            }
        }
        return obj;
    }

    private static object GetValue_Imp(object source, string name)
    {
        if (source == null)
            return null;
        var type = source.GetType();

        while (type != null)
        {
            var f = type.GetField(name, BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance);
            if (f != null)
                return f.GetValue(source);

            var p = type.GetProperty(name, BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance | BindingFlags.IgnoreCase);
            if (p != null)
                return p.GetValue(source, null);

            type = type.BaseType;
        }
        return null;
    }

    private static object GetValue_Imp(object source, string name, int index)
    {
        var enumerable = GetValue_Imp(source, name) as System.Collections.IEnumerable;
        if (enumerable == null) return null;
        var enm = enumerable.GetEnumerator();

        for (int i = 0; i <= index; i++)
        {
            if (!enm.MoveNext()) return null;
        }
        return enm.Current;
    }
}
```

You should be able to just check isArray to draw the struct differently based on whether it's inside a List or not which is likely very handy for you.

I've not looked at Kybernetiks implementation but it ought to be very similar (but he hinted that its more optimized).

Lastly I'd just like to add that sometimes its just easier to serialize something explicitly just instead of fighting with Unitys shortcomings. Unless you're optimizing heavily for memory so you can't have the string as a serialized struct member I'd just serialize it and call it a day.

Btw you were previously also interested in serializing readonly fields earlier and I did stumble over this bit in the Experimental Scripting Previews forum that could be of interest:

https://forum.unity.com/threads/c-7-3-field-serializefield-support.573988/

**Comment by upscalebaby** [comment, marked as answer]
@jasonlawless Okay, so I just randomly mashed my keyboard late tonight and I accidently discovered this:

```cs
public override void OnGUI(Rect position, SerializedProperty property, GUIContent label)
{
    var someProperty = property.FindPropertyRelative("someProperty");
    var serializedObject = new SerializedObject(someProperty.objectReferenceValue);
    var targetObject = serializedObject.targetObject;
}
```

I feel like I just rediscovered atlantis or something. I haven't looked deeper into this but I have verified you do actually get the correct target object instance by doing this!
